File: dinic/dinic.py
import collections

class Edge:
    """Класс для представления ребра графа."""
    def __init__(self, to, capacity):
        self.to = to
        self.capacity = capacity
        self.flow = 0
        self.rev = None

class Dinic:
    """Класс для реализации алгоритма Диница."""
    def __init__(self, n):
        self.n = n
        self.graph = [[] for _ in range(n)]
    
    def add_edge(self, u, v, capacity):
        """Добавляет ребро с пропускной способностью."""
        forward = Edge(v, capacity)
        backward = Edge(u, 0)
        forward.rev = backward
        backward.rev = forward
        self.graph[u].append(forward)
        self.graph[v].append(backward)
    
    def bfs(self, s, t):
        """Построение уровневого графа (обход в ширину)."""
        self.level = [-1] * self.n
        self.level[s] = 0
        queue = collections.deque([s])
        while queue:
            u = queue.popleft()
            for edge in self.graph[u]:
                if edge.capacity - edge.flow > 0 and self.level[edge.to] == -1:
                    self.level[edge.to] = self.level[u] + 1
                    queue.append(edge.to)
        return self.level[t] != -1
    
    def dfs(self, u, t, flow):
        """Поиск блокирующего потока (обход в глубину)."""
        if u == t:
            return flow
        for edge in self.ptr[u]:
            if edge.capacity - edge.flow > 0 and self.level[edge.to] == self.level[u] + 1:
                pushed = self.dfs(edge.to, t, min(flow, edge.capacity - edge.flow))
                if pushed > 0:
                    edge.flow += pushed
                    edge.rev.flow -= pushed
                    return pushed
        return 0
    
    def max_flow(self, s, t):
        """Вычисление максимального потока."""
        flow = 0
        while self.bfs(s, t):
            self.ptr = [iter(self.graph[u]) for u in range(self.n)]
            while True:
                pushed = self.dfs(s, t, float('inf'))
                if pushed == 0:
                    break
                flow += pushed
        return flow

def main():
    """Пример использования алгоритма Диница."""
    n, m = map(int, input().split())  # Количество вершин и рёбер
    dinic = Dinic(n)
    for _ in range(m):
        u, v, c = map(int, input().split())
        dinic.add_edge(u, v, c)
    s, t = map(int, input().split())  # Исток и сток
    print("Максимальный поток:", dinic.max_flow(s, t))

if __name__ == "__main__":
    main()

File: dinic/dinic.cpp
#include <iostream>
#include <vector>
#include <queue>
#include <climits>
#include <cstring>
using namespace std;

// Класс для представления рёбер
struct Edge {
    int to, rev;
    long long capacity, flow;
    Edge(int to, int rev, long long capacity)
        : to(to), rev(rev), capacity(capacity), flow(0) {}
};

class Dinic {
private:
    vector<vector<Edge>> graph;
    vector<int> level;
    vector<int> ptr;
    int n;

public:
    Dinic(int n) : n(n), graph(n), level(n), ptr(n) {}

    void add_edge(int u, int v, long long capacity) {
        graph[u].emplace_back(v, graph[v].size(), capacity);
        graph[v].emplace_back(u, graph[u].size() - 1, 0); // Обратное ребро
    }

    bool bfs(int s, int t) {
        fill(level.begin(), level.end(), -1);
        queue<int> q;
        q.push(s);
        level[s] = 0;
        while (!q.empty()) {
            int u = q.front();
            q.pop();
            for (const Edge& edge : graph[u]) {
                if (edge.capacity - edge.flow > 0 && level[edge.to] == -1) {
                    level[edge.to] = level[u] + 1;
                    q.push(edge.to);
                }
            }
        }
        return level[t] != -1;
    }

    long long dfs(int u, int t, long long pushed) {
        if (u == t || pushed == 0) {
            return pushed;
        }
        for (int& cid = ptr[u]; cid < graph[u].size(); ++cid) {
            Edge& edge = graph[u][cid];
            if (level[edge.to] != level[u] + 1 || edge.capacity - edge.flow <= 0) {
                continue;
            }
            long long tr = dfs(edge.to, t, min(pushed, edge.capacity - edge.flow));
            if (tr > 0) {
                edge.flow += tr;
                graph[edge.to][edge.rev].flow -= tr;
                return tr;
            }
        }
        return 0;
    }

    long long max_flow(int s, int t) {
        long long flow = 0;
        while (bfs(s, t)) {
            fill(ptr.begin(), ptr.end(), 0);
            while (long long pushed = dfs(s, t, LLONG_MAX)) {
                flow += pushed;
            }
        }
        return flow;
    }
};

#ifndef TESTING // Условная компиляция для отключения main()
int main() {
    int n, m;
    cin >> n >> m; // Чтение количества вершин и рёбер
    Dinic dinic(n);

    for (int i = 0; i < m; ++i) {
        int u, v;
        long long capacity;
        cin >> u >> v >> capacity; // Чтение рёбер
        dinic.add_edge(u, v, capacity);
    }

    int s, t;
    cin >> s >> t; // Чтение истока и стока

    cout << "Maximum Flow: " << dinic.max_flow(s, t) << endl;

    return 0;
}
#endif